

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Deciphering gene regulation through vector fields analysis in pancreatic endocrinogenesis &mdash; cellDancer  documentation</title>
  

  
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/jquery.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/underscore.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    
    <script type="text/javascript" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Decoding the human embryo glutamatergic neurogenesis" href="case_study_hgforebrian.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> cellDancer
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">cellDancer</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_preprocessing.html">Data Preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="case_study_gastrulation.html">Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation</a></li>
<li class="toctree-l1"><a class="reference internal" href="case_study_neuro.html">Inferring RNA velocities on each branch for branching genes in the mouse hippocampus development</a></li>
<li class="toctree-l1"><a class="reference internal" href="case_study_pancreas.html">Deciphering cell identity using the cell-specific reaction rates in pancreatic endocrinogenesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="case_study_hgforebrian.html">Decoding the human embryo glutamatergic neurogenesis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application of dynamo</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Deciphering gene regulation through vector fields analysis in pancreatic endocrinogenesis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Import-packages">Import packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Input-the-prediction-of-cellDancer-to-dynamo">Input the prediction of cellDancer to dynamo</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Project-the-RNA-velocity-onto-the-embedding-space">Project the RNA velocity onto the embedding space</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Learn-the-vector-fields">Learn the vector fields</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Jacobian-analysis">Jacobian analysis</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cellDancer</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Deciphering gene regulation through vector fields analysis in pancreatic endocrinogenesis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/case_study_pancreas_dynamo.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Deciphering-gene-regulation-through-vector-fields-analysis-in-pancreatic-endocrinogenesis">
<h1>Deciphering gene regulation through vector fields analysis in pancreatic endocrinogenesis<a class="headerlink" href="#Deciphering-gene-regulation-through-vector-fields-analysis-in-pancreatic-endocrinogenesis" title="Permalink to this headline">¶</a></h1>
<p>cellDancer could be integrated into downstream analysis, such as <a class="reference external" href="https://dynamo-release.readthedocs.io/en/latest/index.html">dynamo</a> proposed by <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0092867421015774#!">Qiu et al.</a></p>
<p>In this tutorial, the output of cellDancer (<a class="reference external" href="https://drive.google.com/file/d/1n_ZxAXuKB1BzcpIDkPj3XdOs-Ni45yD7/view?usp=sharing">PancreaticEndocrinogenesis_cellDancer_estimation.csv</a>) in pancreatic endocrinogenesis is transferred to adata as dynamo input. We apply dynamo to (1) perform vector field analysis and (2) detect gene regulation with Jacobian analysis.</p>
<div class="section" id="Import-packages">
<h2>Import packages<a class="headerlink" href="#Import-packages" title="Permalink to this headline">¶</a></h2>
<p>To run the notebook locally, both cellDancer and dynamo are need. Refer <a class="reference external" href="installation.html">cellDancer</a> and <a class="reference external" href="https://dynamo-release.readthedocs.io/en/latest/index.html">dynamo</a> for instructions of building the packages.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">dynamo</span> <span class="k">as</span> <span class="nn">dyn</span>
<span class="kn">from</span> <span class="nn">celldancer.utilities</span> <span class="kn">import</span> <span class="n">export_velocity_to_dynamo</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; setting visualization default mode in dynamo. Your customized matplotlib settings might be overritten.
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dyn</span><span class="o">.</span><span class="n">configuration</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="s1">&#39;dynamo&#39;</span><span class="p">,</span> <span class="n">background</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Input-the-prediction-of-cellDancer-to-dynamo">
<h2>Input the prediction of cellDancer to dynamo<a class="headerlink" href="#Input-the-prediction-of-cellDancer-to-dynamo" title="Permalink to this headline">¶</a></h2>
<p>We import the cellDancer predicted RNA velocity to dynamo adata with API <code class="docutils literal notranslate"><span class="pre">export_velocity_to_dynamo()</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># First create dynamo adata</span>
<span class="n">adata_dyn</span> <span class="o">=</span> <span class="n">dyn</span><span class="o">.</span><span class="n">sample_data</span><span class="o">.</span><span class="n">pancreatic_endocrinogenesis</span><span class="p">()</span>
<span class="n">pancreas_genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Hes1&quot;</span><span class="p">,</span><span class="s2">&quot;Nkx6-1&quot;</span><span class="p">,</span><span class="s2">&quot;Nkx2-2&quot;</span><span class="p">,</span><span class="s2">&quot;Neurog3&quot;</span><span class="p">,</span><span class="s2">&quot;Neurod1&quot;</span><span class="p">,</span><span class="s2">&quot;Pax4&quot;</span><span class="p">,</span><span class="s2">&quot;Pax6&quot;</span><span class="p">,</span><span class="s2">&quot;Arx&quot;</span><span class="p">,</span><span class="s2">&quot;Pdx1&quot;</span><span class="p">,</span><span class="s2">&quot;Ins1&quot;</span><span class="p">,</span><span class="s2">&quot;Ins2&quot;</span><span class="p">,</span><span class="s2">&quot;Ghrl&quot;</span><span class="p">,</span><span class="s2">&quot;Ptf1a&quot;</span><span class="p">,</span><span class="s2">&quot;Iapp&quot;</span><span class="p">,</span><span class="s2">&quot;Isl1&quot;</span><span class="p">,</span> <span class="s2">&quot;Sox9&quot;</span><span class="p">,</span><span class="s2">&quot;Gcg&quot;</span><span class="p">]</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">recipe_monocle</span><span class="p">(</span><span class="n">adata_dyn</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span> <span class="n">fg_kwargs</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;shared_count&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">},</span> <span class="n">genes_to_append</span><span class="o">=</span><span class="n">pancreas_genes</span><span class="p">)</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">dynamics</span><span class="p">(</span><span class="n">adata_dyn</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s1">&#39;static&#39;</span><span class="p">)</span>

<span class="c1"># replace adata.layers[&#39;velocity_S&#39;], adata.var[&#39;use_for_dynamics&#39;], and adata.var[&#39;use_for_transition&#39;] with cellDancer outputs.</span>
<span class="n">cellDancer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;your_path/PancreaticEndocrinogenesis_cellDancer_estimation.csv&#39;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">export_velocity_to_dynamo</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="p">,</span><span class="n">adata_dyn</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; Downloading data to ./data/endocrinogenesis_day15.h5ad
|-----&gt; recipe_monocle_keep_filtered_cells_key is None. Using default value from DynamoAdataConfig: recipe_monocle_keep_filtered_cells_key=True
|-----&gt; recipe_monocle_keep_filtered_genes_key is None. Using default value from DynamoAdataConfig: recipe_monocle_keep_filtered_genes_key=True
|-----&gt; recipe_monocle_keep_raw_layers_key is None. Using default value from DynamoAdataConfig: recipe_monocle_keep_raw_layers_key=True
|-----&gt; apply Monocole recipe to adata...
|-----&gt; &lt;insert&gt; pp to uns in AnnData Object.
|-----------&gt; &lt;insert&gt; has_splicing to uns[&#39;pp&#39;] in AnnData Object.
|-----------&gt; &lt;insert&gt; has_labling to uns[&#39;pp&#39;] in AnnData Object.
|-----------&gt; &lt;insert&gt; splicing_labeling to uns[&#39;pp&#39;] in AnnData Object.
|-----------&gt; &lt;insert&gt; has_protein to uns[&#39;pp&#39;] in AnnData Object.
|-----&gt; ensure all cell and variable names unique.
|-----&gt; ensure all data in different layers in csr sparse matrix format.
|-----&gt; ensure all labeling data properly collapased
|-----------&gt; &lt;insert&gt; tkey to uns[&#39;pp&#39;] in AnnData Object.
|-----------&gt; &lt;insert&gt; experiment_type to uns[&#39;pp&#39;] in AnnData Object.
|-----&gt; filtering cells...
|-----&gt; &lt;insert&gt; pass_basic_filter to obs in AnnData Object.
|-----&gt; 3696 cells passed basic filters.
|-----&gt; filtering gene...
|-----&gt; &lt;insert&gt; pass_basic_filter to var in AnnData Object.
|-----&gt; 6956 genes passed basic filters.
|-----&gt; calculating size factor...
|-----&gt; selecting genes in layer: X, sort method: SVR...
|-----&gt; &lt;insert&gt; frac to var in AnnData Object.
|-----&gt; size factor normalizing the data, followed by log1p transformation.
|-----&gt; Set &lt;adata.X&gt; to normalized data
|-----&gt; applying PCA ...
|-----&gt; &lt;insert&gt; pca_fit to uns in AnnData Object.
|-----&gt; &lt;insert&gt; ntr to obs in AnnData Object.
|-----&gt; &lt;insert&gt; ntr to var in AnnData Object.
|-----&gt; cell cycle scoring...
|-----&gt; computing cell phase...
|-----&gt; [cell phase estimation] in progress: 100.0000%
|-----&gt; [cell phase estimation] finished [23.2167s]
|-----&gt; &lt;insert&gt; cell_cycle_phase to obs in AnnData Object.
|-----&gt; &lt;insert&gt; cell_cycle_scores to obsm in AnnData Object.
|-----&gt; [Cell Cycle Scores Estimation] in progress: 100.0000%
|-----&gt; [Cell Cycle Scores Estimation] finished [0.5055s]
|-----&gt; [recipe_monocle preprocess] in progress: 100.0000%
|-----&gt; [recipe_monocle preprocess] finished [7.7637s]
|-----&gt; dynamics_del_2nd_moments_key is None. Using default value from DynamoAdataConfig: dynamics_del_2nd_moments_key=False
|-----------&gt; removing existing M layers:[]...
|-----------&gt; making adata smooth...
|-----&gt; calculating first/second moments...
|-----&gt; [moments calculation] in progress: 100.0000%
|-----&gt; [moments calculation] finished [19.5928s]
</pre></div></div>
</div>
</div>
<div class="section" id="Project-the-RNA-velocity-onto-the-embedding-space">
<h2>Project the RNA velocity onto the embedding space<a class="headerlink" href="#Project-the-RNA-velocity-onto-the-embedding-space" title="Permalink to this headline">¶</a></h2>
<p>The projection of RNA velocity imported from cellDancer onto the embedding space was done by using dynamo method <code class="docutils literal notranslate"><span class="pre">cell_velocities()</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pancreas_cluster_cmap</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Ductal&#39;</span><span class="p">:</span> <span class="s1">&#39;#3361A5&#39;</span><span class="p">,</span><span class="s1">&#39;Ngn3 low EP&#39;</span><span class="p">:</span> <span class="s1">&#39;#248AF3&#39;</span><span class="p">,</span><span class="s1">&#39;Ngn3 high EP&#39;</span><span class="p">:</span> <span class="s1">&#39;#14B3FF&#39;</span><span class="p">,</span><span class="s1">&#39;Pre-endocrine&#39;</span><span class="p">:</span> <span class="s1">&#39;#88CEEF&#39;</span><span class="p">,</span><span class="s1">&#39;Alpha&#39;</span><span class="p">:</span> <span class="s1">&#39;#ff4800&#39;</span><span class="p">,</span><span class="s1">&#39;Beta&#39;</span><span class="p">:</span> <span class="s1">&#39;#B81136&#39;</span><span class="p">,</span><span class="s1">&#39;Delta&#39;</span><span class="p">:</span> <span class="s1">&#39;green&#39;</span><span class="p">,</span><span class="s1">&#39;Epsilon&#39;</span><span class="p">:</span> <span class="s1">&#39;#03B3B0&#39;</span><span class="p">}</span>

<span class="n">dyn</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">reduceDimension</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pca_components</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">cell_velocities</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;pearson&quot;</span><span class="p">,</span> <span class="n">other_kernels_dict</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;transform&quot;</span><span class="p">:</span> <span class="s2">&quot;sqrt&quot;</span><span class="p">})</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">cell_velocities</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>

<span class="n">dyn</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">streamline_plot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;clusters&quot;</span><span class="p">],</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">,</span> <span class="n">color_key</span> <span class="o">=</span> <span class="n">pancreas_cluster_cmap</span><span class="p">,</span> <span class="n">show_legend</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">,</span> <span class="n">show_arrowed_spines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; retrive data for non-linear dimension reduction...
|-----? adata already have basis umap. dimension reduction umap will be skipped!
set enforce=True to re-performing dimension reduction.
|-----&gt; [dimension_reduction projection] in progress: 100.0000%
|-----&gt; [dimension_reduction projection] finished [0.0039s]
|-----&gt; 0 genes are removed because of nan velocity values.
|-----&gt; [calculating transition matrix via pearson kernel with sqrt transform.] in progress: 100.0000%
|-----&gt; [calculating transition matrix via pearson kernel with sqrt transform.] finished [14.3924s]
|-----&gt; [projecting velocity vector to low dimensional embedding] in progress: 100.0000%
|-----&gt; [projecting velocity vector to low dimensional embedding] finished [1.0276s]
|-----&gt; 0 genes are removed because of nan velocity values.
Using existing pearson_transition_matrix found in .obsp.
|-----&gt; [projecting velocity vector to low dimensional embedding] in progress: 100.0000%
|-----&gt; [projecting velocity vector to low dimensional embedding] finished [1.0609s]
|-----------&gt; plotting with basis key=X_umap
|-----------&gt; skip filtering clusters by stack threshold when stacking color because it is not a numeric type
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_pancreas_dynamo_11_1.png" class="no-scaled-link" src="../_images/notebooks_case_study_pancreas_dynamo_11_1.png" style="width: 487px; height: 340px;" />
</div>
</div>
</div>
<div class="section" id="Learn-the-vector-fields">
<h2>Learn the vector fields<a class="headerlink" href="#Learn-the-vector-fields" title="Permalink to this headline">¶</a></h2>
<p>We use dynamo to learn and visualize the vector fields on UMAP.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dyn</span><span class="o">.</span><span class="n">vf</span><span class="o">.</span><span class="n">VectorField</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;pca&quot;</span><span class="p">,</span> <span class="n">pot_curl_div</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">vf</span><span class="o">.</span><span class="n">VectorField</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">,</span> <span class="n">pot_curl_div</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">dyn</span><span class="o">.</span><span class="n">vf</span><span class="o">.</span><span class="n">speed</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">vf</span><span class="o">.</span><span class="n">divergence</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">vf</span><span class="o">.</span><span class="n">acceleration</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="n">dyn</span><span class="o">.</span><span class="n">vf</span><span class="o">.</span><span class="n">curl</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; VectorField reconstruction begins...
|-----&gt; Retrieve X and V based on basis: PCA.
        Vector field will be learned in the PCA space.
|-----&gt; Learning vector field with method: sparsevfc.
|-----&gt; [SparseVFC] begins...
|-----&gt; Sampling control points based on data velocity magnitude...
|-----&gt; [SparseVFC] in progress: 100.0000%
|-----&gt; [SparseVFC] finished [0.3184s]
|-----&gt; &lt;insert&gt; velocity_pca_SparseVFC to obsm in AnnData Object.
|-----&gt; &lt;insert&gt; X_pca_SparseVFC to obsm in AnnData Object.
|-----&gt; &lt;insert&gt; VecFld_pca to uns in AnnData Object.
|-----&gt; Running ddhodge to estimate vector field based pseudotime in pca basis...
|-----&gt; graphizing vectorfield...
|-----&gt; incomplete neighbor graph info detected: connectivities and distances do not exist in adata.obsp, indices not in adata.uns.neighbors.
|-----&gt; Neighbor graph is broken, recomputing....
|-----&gt; Start computing neighbor graph...
|-----------&gt; X_data is None, fetching or recomputing...
|-----&gt; fetching X data from layer:None, basis:pca
|-----&gt; method arg is None, choosing methods automatically...
|-----------&gt; method ball_tree selected
|-----&gt; &lt;insert&gt; connectivities to obsp in AnnData Object.
|-----&gt; &lt;insert&gt; distances to obsp in AnnData Object.
|-----&gt; &lt;insert&gt; neighbors to uns in AnnData Object.
|-----&gt; &lt;insert&gt; neighbors.indices to uns in AnnData Object.
|-----&gt; &lt;insert&gt; neighbors.params to uns in AnnData Object.
|-----------? nbrs_idx argument is ignored and recomputed because nbrs_idx is not None and return_nbrs=True
|-----------&gt; calculating neighbor indices...
|-----&gt; [ddhodge completed] in progress: 100.0000%
|-----&gt; [ddhodge completed] finished [20.5809s]
|-----&gt; Computing divergence...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating divergence: 100%|█████████████████████| 4/4 [00:00&lt;00:00,  9.94it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; &lt;insert&gt; control_point_pca to obs in AnnData Object.
|-----&gt; &lt;insert&gt; inlier_prob_pca to obs in AnnData Object.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; &lt;insert&gt; obs_vf_angle_pca to obs in AnnData Object.
|-----&gt; [VectorField] in progress: 100.0000%
|-----&gt; [VectorField] finished [21.5317s]
|-----&gt; VectorField reconstruction begins...
|-----&gt; Retrieve X and V based on basis: UMAP.
        Vector field will be learned in the UMAP space.
|-----&gt; Generating high dimensional grids and convert into a row matrix.
|-----&gt; Learning vector field with method: sparsevfc.
|-----&gt; [SparseVFC] begins...
|-----&gt; Sampling control points based on data velocity magnitude...
|-----&gt; [SparseVFC] in progress: 100.0000%
|-----&gt; [SparseVFC] finished [0.4374s]
|-----&gt; &lt;insert&gt; velocity_umap_SparseVFC to obsm in AnnData Object.
|-----&gt; &lt;insert&gt; X_umap_SparseVFC to obsm in AnnData Object.
|-----&gt; &lt;insert&gt; VecFld_umap to uns in AnnData Object.
|-----&gt; Running ddhodge to estimate vector field based pseudotime in umap basis...
|-----&gt; graphizing vectorfield...
|-----------? nbrs_idx argument is ignored and recomputed because nbrs_idx is not None and return_nbrs=True
|-----------&gt; calculating neighbor indices...
|-----&gt; [ddhodge completed] in progress: 100.0000%
|-----&gt; [ddhodge completed] finished [18.7982s]
|-----&gt; Computing curl...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating 2-D curl: 100%|██████████████| 3696/3696 [00:00&lt;00:00, 15543.98it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; Computing divergence...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

Calculating divergence: 100%|█████████████████████| 4/4 [00:00&lt;00:00, 23.67it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; &lt;insert&gt; control_point_umap to obs in AnnData Object.
|-----&gt; &lt;insert&gt; inlier_prob_umap to obs in AnnData Object.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; &lt;insert&gt; obs_vf_angle_umap to obs in AnnData Object.
|-----&gt; [VectorField] in progress: 100.0000%
|-----&gt; [VectorField] finished [19.8864s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating divergence: 100%|█████████████████████| 4/4 [00:00&lt;00:00, 10.07it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; [Calculating acceleration] in progress: 100.0000%
|-----&gt; [Calculating acceleration] finished [0.1751s]
|-----&gt; &lt;insert&gt; acceleration to layers in AnnData Object.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating 2-D curl: 100%|██████████████| 3696/3696 [00:00&lt;00:00, 17567.05it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dyn</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">topography</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">],</span> <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">background</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span>
                  <span class="n">streamline_color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">color_key</span> <span class="o">=</span> <span class="n">pancreas_cluster_cmap</span><span class="p">,</span> <span class="n">show_legend</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">terms</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;streamline&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----&gt; Vector field for umap is but its topography is not mapped. Mapping topography now ...
|-----------&gt; plotting with basis key=X_umap
|-----------&gt; skip filtering clusters by stack threshold when stacking color because it is not a numeric type
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_pancreas_dynamo_15_1.png" class="no-scaled-link" src="../_images/notebooks_case_study_pancreas_dynamo_15_1.png" style="width: 590px; height: 389px;" />
</div>
</div>
<p>The emitting fixed point #0 is identified in ductal cells. The absorbing fixed points #1 #2 #3 are identified in Alpha-, Beta-, and Epsilon-cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_load_fix_points</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">Xss</span><span class="p">,</span> <span class="n">ftype</span><span class="p">,</span> <span class="n">conf</span> <span class="o">=</span> <span class="n">adata_load_fix_points</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;VecFld_umap&#39;</span><span class="p">][</span><span class="s1">&#39;Xss&#39;</span><span class="p">],</span>\
                   <span class="n">adata_load_fix_points</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;VecFld_umap&#39;</span><span class="p">][</span><span class="s1">&#39;ftype&#39;</span><span class="p">],</span>\
                   <span class="n">adata_load_fix_points</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;VecFld_umap&#39;</span><span class="p">][</span><span class="s1">&#39;confidence&#39;</span><span class="p">]</span>

<span class="n">fixed_points</span> <span class="o">=</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">16</span><span class="p">,</span><span class="mi">12</span><span class="p">]</span>

<span class="n">adata_load_fix_points</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;VecFld_umap&#39;</span><span class="p">][</span><span class="s1">&#39;Xss&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Xss</span><span class="p">[</span><span class="n">fixed_points</span><span class="p">]</span>
<span class="n">adata_load_fix_points</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;VecFld_umap&#39;</span><span class="p">][</span><span class="s1">&#39;ftype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftype</span><span class="p">[</span><span class="n">fixed_points</span><span class="p">]</span>
<span class="n">adata_load_fix_points</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;VecFld_umap&#39;</span><span class="p">][</span><span class="s1">&#39;confidence&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">conf</span><span class="p">[</span><span class="n">fixed_points</span><span class="p">]</span>

<span class="n">dyn</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">topography</span><span class="p">(</span>
    <span class="n">adata_load_fix_points</span><span class="p">,</span>
    <span class="n">markersize</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">,</span>
    <span class="n">fps_basis</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">,</span>
    <span class="n">color_key</span> <span class="o">=</span> <span class="n">pancreas_cluster_cmap</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;clusters&quot;</span><span class="p">],</span>
    <span class="n">show_legend</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">,</span>
    <span class="n">save_show_or_return</span><span class="o">=</span><span class="s1">&#39;show&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----------&gt; plotting with basis key=X_umap
|-----------&gt; skip filtering clusters by stack threshold when stacking color because it is not a numeric type
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_pancreas_dynamo_17_1.png" class="no-scaled-link" src="../_images/notebooks_case_study_pancreas_dynamo_17_1.png" style="width: 590px; height: 389px;" />
</div>
</div>
</div>
<div class="section" id="Jacobian-analysis">
<h2>Jacobian analysis<a class="headerlink" href="#Jacobian-analysis" title="Permalink to this headline">¶</a></h2>
<p>Jacobian analysis shows that Pax4 is down-regulated by Arx in the Alpha-cells; Arx is down-regulated by Pax4 in the Beta-cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dyn</span><span class="o">.</span><span class="n">vf</span><span class="o">.</span><span class="n">jacobian</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">regulators</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Arx&#39;</span><span class="p">,</span><span class="s1">&#39;Pax4&#39;</span><span class="p">])</span>

<span class="n">dyn</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">jacobian</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">,</span>
    <span class="n">regulators</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Arx&#39;</span><span class="p">],</span>
    <span class="n">effectors</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pax4&#39;</span><span class="p">],</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">dyn</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">jacobian</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">,</span>
    <span class="n">regulators</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pax4&#39;</span><span class="p">],</span>
    <span class="n">effectors</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Arx&#39;</span><span class="p">],</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Transforming subset Jacobian: 100%|██████| 3696/3696 [00:00&lt;00:00, 98024.28it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_pancreas_dynamo_20_1.png" class="no-scaled-link" src="../_images/notebooks_case_study_pancreas_dynamo_20_1.png" style="width: 487px; height: 360px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_pancreas_dynamo_20_2.png" class="no-scaled-link" src="../_images/notebooks_case_study_pancreas_dynamo_20_2.png" style="width: 487px; height: 360px;" />
</div>
</div>
<p>We use dynamo to plot the expression of Arx and Pax4 on UMAP.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dyn</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>  <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Arx&quot;</span><span class="p">,</span><span class="s1">&#39;Pax4&#39;</span><span class="p">],</span><span class="n">save_show_or_return</span><span class="o">=</span><span class="s1">&#39;show&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
|-----------&gt; plotting with basis key=X_umap
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_pancreas_dynamo_22_1.png" class="no-scaled-link" src="../_images/notebooks_case_study_pancreas_dynamo_22_1.png" style="width: 889px; height: 350px;" />
</div>
</div>
<script type="application/vnd.jupyter.widget-state+json">
{"state": {}, "version_major": 2, "version_minor": 0}
</script></div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="case_study_hgforebrian.html" class="btn btn-neutral float-left" title="Decoding the human embryo glutamatergic neurogenesis" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, 6.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>