

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation &mdash; cellDancer  documentation</title>
  

  
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/jquery.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/underscore.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    
    <script type="text/javascript" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Inferring RNA velocities on each branch for branching genes in the mouse hippocampus development" href="case_study_neuro.html" />
    <link rel="prev" title="Data Preparation" href="../data_preprocessing.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> cellDancer
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">cellDancer</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_preprocessing.html">Data Preparation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Import-packages">Import packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Load-data">Load data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Estimate-RNA-velocity-for-sample-genes">Estimate RNA velocity for sample genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Visualize-the-phase-portraits">Visualize the phase portraits</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Project-the-RNA-velocity-onto-the-embedding-space">Project the RNA velocity onto the embedding space</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Estimate-pseudotime">Estimate pseudotime</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Display-the-abundance-of-spliced-RNA-along-pseudotime">Display the abundance of spliced RNA along pseudotime</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="case_study_neuro.html">Inferring RNA velocities on each branch for branching genes in the mouse hippocampus development</a></li>
<li class="toctree-l1"><a class="reference internal" href="case_study_pancreas.html">Deciphering cell identity using the cell-specific reaction rates in pancreatic endocrinogenesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="case_study_hgforebrian.html">Decoding the human embryo glutamatergic neurogenesis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application of dynamo</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="case_study_pancreas_dynamo.html">Deciphering gene regulation through vector fields analysis in pancreatic endocrinogenesis</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cellDancer</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/case_study_gastrulation.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Delineating-the-transcriptional-boost-on-single-cell-resolution-in-gastrulation-erythroid-maturation">
<h1>Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation<a class="headerlink" href="#Delineating-the-transcriptional-boost-on-single-cell-resolution-in-gastrulation-erythroid-maturation" title="Permalink to this headline">¶</a></h1>
<p>This tutorial shows how cellDancer analyzes RNA velocity. The process includes (1) estimating RNA velocity, (2) deriving cell fates on embedding space, and (3) estimating pseudotime.</p>
<p>This tutorial also demonstrates cellDancer’s ability to resolve the difficulty of predicting the velocity of genes with transcriptional boost mentioned by <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02414-y">Bergen et al.</a></p>
<p>Below is the case study for gastrulation erythroid maturation. We select 12,329 cells with 2,000 genes from haemato-endothelial progenitors, blood progenitors 1/2, and erythroid 1/2/3. The embryonic days are selected from E7.0, E7.25, E7.5, E7.75, E8.0, E8.25, and E8.5.</p>
<div class="section" id="Import-packages">
<h2>Import packages<a class="headerlink" href="#Import-packages" title="Permalink to this headline">¶</a></h2>
<p>To run the notebook locally, <a class="reference external" href="installation.html">Installation</a> could be referred to install the environment and dependencies.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">celldancer</span> <span class="k">as</span> <span class="nn">cd</span>
<span class="kn">import</span> <span class="nn">celldancer.cdplt</span> <span class="k">as</span> <span class="nn">cdplt</span>
<span class="kn">from</span> <span class="nn">celldancer.cdplt</span> <span class="kn">import</span> <span class="n">colormap</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Permalink to this headline">¶</a></h2>
<p>The input data for cellDancer contains the abundances of unspliced RNA and spliced RNA. For the detail of obtaining the dataset and pre-processing steps, <a class="reference external" href="data_preprocessing.html">Data Preparation</a> could be referred to. We follow the standard data preparation procedures in <a class="reference external" href="https://scvelo.readthedocs.io/">scVelo</a> with default parameters except that we use 100 nearest neighbors for first-moment calculation.</p>
<p>The data of gastrulation erythroid maturation could be downloaded and unzipped from <a class="reference external" href="https://drive.google.com/file/d/1DSbxJr5dh-NJp4Bx20tNlZHSjGjyigw-/view?usp=sharing">GastrulationErythroid_cell_type_u_s.csv.zip</a>. It could be loaded by <code class="docutils literal notranslate"><span class="pre">pd.read_csv('your_path/GastrulationErythroid_cell_type_u_s.csv')</span></code>. To load your own data, the dataframe should contain columns ‘gene_name’, ‘unsplice’, ‘splice’ ,‘cellID’ ,‘clusters’ ,‘embedding1’ , and ‘embedding2.’ For a detailed description of the data
structure, <a class="reference external" href="data_preprocessing.html">Preprocessing</a> could be referred to.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_type_u_s_path</span><span class="o">=</span><span class="s1">&#39;your_path/GastrulationErythroid_cell_type_u_s.csv&#39;</span>
<span class="n">cell_type_u_s</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cell_type_u_s_path</span><span class="p">)</span>
<span class="n">cell_type_u_s</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_name</th>
      <th>unsplice</th>
      <th>splice</th>
      <th>cellID</th>
      <th>clusters</th>
      <th>embedding1</th>
      <th>embedding2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Sox17</td>
      <td>0.000000</td>
      <td>0.043971</td>
      <td>cell_363</td>
      <td>Blood progenitors 2</td>
      <td>3.460521</td>
      <td>15.574629</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Sox17</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>cell_382</td>
      <td>Blood progenitors 2</td>
      <td>2.490433</td>
      <td>14.971734</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Sox17</td>
      <td>0.000000</td>
      <td>0.018161</td>
      <td>cell_385</td>
      <td>Blood progenitors 2</td>
      <td>2.351203</td>
      <td>15.267069</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Sox17</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>cell_393</td>
      <td>Blood progenitors 2</td>
      <td>5.899098</td>
      <td>14.388825</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Sox17</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>cell_398</td>
      <td>Blood progenitors 2</td>
      <td>4.823139</td>
      <td>15.374831</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>24657995</th>
      <td>Gm47283</td>
      <td>0.214961</td>
      <td>1.145533</td>
      <td>cell_139318</td>
      <td>Erythroid3</td>
      <td>8.032358</td>
      <td>7.603037</td>
    </tr>
    <tr>
      <th>24657996</th>
      <td>Gm47283</td>
      <td>0.300111</td>
      <td>1.072944</td>
      <td>cell_139321</td>
      <td>Erythroid3</td>
      <td>10.352904</td>
      <td>6.446736</td>
    </tr>
    <tr>
      <th>24657997</th>
      <td>Gm47283</td>
      <td>0.292607</td>
      <td>1.199875</td>
      <td>cell_139326</td>
      <td>Erythroid3</td>
      <td>9.464873</td>
      <td>7.261099</td>
    </tr>
    <tr>
      <th>24657998</th>
      <td>Gm47283</td>
      <td>0.266031</td>
      <td>1.114659</td>
      <td>cell_139327</td>
      <td>Erythroid3</td>
      <td>9.990495</td>
      <td>7.243880</td>
    </tr>
    <tr>
      <th>24657999</th>
      <td>Gm47283</td>
      <td>0.164675</td>
      <td>1.111358</td>
      <td>cell_139330</td>
      <td>Erythroid3</td>
      <td>8.260699</td>
      <td>7.935455</td>
    </tr>
  </tbody>
</table>
<p>24658000 rows × 7 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Estimate-RNA-velocity-for-sample-genes">
<h2>Estimate RNA velocity for sample genes<a class="headerlink" href="#Estimate-RNA-velocity-for-sample-genes" title="Permalink to this headline">¶</a></h2>
<p>Here, 15 genes in <code class="docutils literal notranslate"><span class="pre">gene_list</span></code> are estimated as an example. The predicted unspliced and spliced reads, alpha, beta, and gamma are added to the dataframe.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Smarca2&#39;</span><span class="p">,</span> <span class="s1">&#39;Rbms2&#39;</span><span class="p">,</span> <span class="s1">&#39;Myo1b&#39;</span><span class="p">,</span> <span class="s1">&#39;Hba-x&#39;</span><span class="p">,</span> <span class="s1">&#39;Yipf5&#39;</span><span class="p">,</span> <span class="s1">&#39;Skap1&#39;</span><span class="p">,</span> <span class="s1">&#39;Smim1&#39;</span><span class="p">,</span> <span class="s1">&#39;Nfkb1&#39;</span><span class="p">,</span> <span class="s1">&#39;Sulf2&#39;</span><span class="p">,</span> <span class="s1">&#39;Blvrb&#39;</span><span class="p">,</span> <span class="s1">&#39;Hbb-y&#39;</span><span class="p">,</span> <span class="s1">&#39;Coro2b&#39;</span><span class="p">,</span> <span class="s1">&#39;Yipf5&#39;</span><span class="p">,</span> <span class="s1">&#39;Phc2&#39;</span><span class="p">,</span> <span class="s1">&#39;Mllt3&#39;</span><span class="p">]</span>

<span class="n">loss_df</span><span class="p">,</span> <span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">velocity</span><span class="p">(</span><span class="n">cell_type_u_s</span><span class="p">,</span>\
                                   <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>\
                                   <span class="n">permutation_ratio</span><span class="o">=</span><span class="mf">0.125</span><span class="p">,</span>\
                                   <span class="n">n_jobs</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">cellDancer_df</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Using /Users/shengyuli/Library/CloudStorage/OneDrive-HoustonMethodist/work/Velocity/bin/cellDancer_polish/analysis/CaseStudyNotebook/cellDancer_velocity_2022-10-20 10-11-23 as the output path.
Arranging genes for parallel job.
14  genes were arranged to  2  portions.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

Velocity Estimation:   0%|          | 0/2 [00:00&lt;?, ?it/s]
Velocity Estimation:  50%|█████     | 1/2 [00:20&lt;00:20, 20.31s/it]
Velocity Estimation: 100%|██████████| 2/2 [00:29&lt;00:00, 13.99s/it]

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cellIndex</th>
      <th>gene_name</th>
      <th>unsplice</th>
      <th>splice</th>
      <th>unsplice_predict</th>
      <th>splice_predict</th>
      <th>alpha</th>
      <th>beta</th>
      <th>gamma</th>
      <th>loss</th>
      <th>cellID</th>
      <th>clusters</th>
      <th>embedding1</th>
      <th>embedding2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>Mllt3</td>
      <td>0.087552</td>
      <td>0.061677</td>
      <td>0.131864</td>
      <td>0.347319</td>
      <td>0.264903</td>
      <td>0.304851</td>
      <td>0.335579</td>
      <td>0.111875</td>
      <td>cell_363</td>
      <td>Blood progenitors 2</td>
      <td>3.460521</td>
      <td>15.574629</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>Mllt3</td>
      <td>0.042049</td>
      <td>0.056451</td>
      <td>0.096471</td>
      <td>0.192440</td>
      <td>0.253070</td>
      <td>0.303873</td>
      <td>0.335266</td>
      <td>0.111875</td>
      <td>cell_382</td>
      <td>Blood progenitors 2</td>
      <td>2.490433</td>
      <td>14.971734</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>Mllt3</td>
      <td>0.046162</td>
      <td>0.073611</td>
      <td>0.099587</td>
      <td>0.222645</td>
      <td>0.253818</td>
      <td>0.304017</td>
      <td>0.335363</td>
      <td>0.111875</td>
      <td>cell_385</td>
      <td>Blood progenitors 2</td>
      <td>2.351203</td>
      <td>15.267069</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>Mllt3</td>
      <td>0.119952</td>
      <td>0.142994</td>
      <td>0.156111</td>
      <td>0.534318</td>
      <td>0.270146</td>
      <td>0.306117</td>
      <td>0.336443</td>
      <td>0.111875</td>
      <td>cell_393</td>
      <td>Blood progenitors 2</td>
      <td>5.899098</td>
      <td>14.388825</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>Mllt3</td>
      <td>0.027221</td>
      <td>0.136220</td>
      <td>0.083502</td>
      <td>0.221573</td>
      <td>0.243896</td>
      <td>0.304314</td>
      <td>0.336083</td>
      <td>0.111875</td>
      <td>cell_398</td>
      <td>Blood progenitors 2</td>
      <td>4.823139</td>
      <td>15.374831</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>172601</th>
      <td>12324</td>
      <td>Yipf5</td>
      <td>0.098889</td>
      <td>0.443361</td>
      <td>0.101194</td>
      <td>0.466415</td>
      <td>0.080456</td>
      <td>0.198707</td>
      <td>0.447713</td>
      <td>0.060657</td>
      <td>cell_139318</td>
      <td>Erythroid3</td>
      <td>8.032358</td>
      <td>7.603037</td>
    </tr>
    <tr>
      <th>172602</th>
      <td>12325</td>
      <td>Yipf5</td>
      <td>0.092605</td>
      <td>0.521375</td>
      <td>0.094589</td>
      <td>0.530523</td>
      <td>0.074335</td>
      <td>0.197698</td>
      <td>0.450130</td>
      <td>0.060657</td>
      <td>cell_139321</td>
      <td>Erythroid3</td>
      <td>10.352904</td>
      <td>6.446736</td>
    </tr>
    <tr>
      <th>172603</th>
      <td>12326</td>
      <td>Yipf5</td>
      <td>0.108367</td>
      <td>0.443695</td>
      <td>0.109323</td>
      <td>0.473431</td>
      <td>0.081982</td>
      <td>0.198220</td>
      <td>0.447828</td>
      <td>0.060657</td>
      <td>cell_139326</td>
      <td>Erythroid3</td>
      <td>9.464873</td>
      <td>7.261099</td>
    </tr>
    <tr>
      <th>172604</th>
      <td>12327</td>
      <td>Yipf5</td>
      <td>0.067904</td>
      <td>0.431785</td>
      <td>0.074884</td>
      <td>0.433994</td>
      <td>0.076296</td>
      <td>0.200459</td>
      <td>0.447011</td>
      <td>0.060657</td>
      <td>cell_139327</td>
      <td>Erythroid3</td>
      <td>9.990495</td>
      <td>7.243880</td>
    </tr>
    <tr>
      <th>172605</th>
      <td>12328</td>
      <td>Yipf5</td>
      <td>0.077812</td>
      <td>0.388213</td>
      <td>0.084044</td>
      <td>0.402742</td>
      <td>0.080739</td>
      <td>0.200674</td>
      <td>0.445712</td>
      <td>0.060657</td>
      <td>cell_139330</td>
      <td>Erythroid3</td>
      <td>8.260699</td>
      <td>7.935455</td>
    </tr>
  </tbody>
</table>
<p>172606 rows × 14 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Visualize-the-phase-portraits">
<h2>Visualize the phase portraits<a class="headerlink" href="#Visualize-the-phase-portraits" title="Permalink to this headline">¶</a></h2>
<p>We visualize the phase portrait of each gene with <code class="docutils literal notranslate"><span class="pre">cdplt.scatter_gene()</span></code>.</p>
<p>From the figures below, the prediction of those genes that have transcriptional boost (Hba-x, Smim1, and Hbb-y) are consistent with the progression of gastrulation erythroid maturation of (1) haemato-endothelial progenitors, (2) blood progenitors 1/2, and (3) erythroid. This reflects the ability of cellDancer to resolve the RNA velocity for the genes in the transcriptional boost scenario.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ncols</span><span class="o">=</span><span class="mi">5</span>
<span class="n">height</span><span class="o">=</span><span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="o">/</span><span class="mi">5</span><span class="p">)</span><span class="o">*</span><span class="mi">4</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="n">height</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="o">/</span><span class="n">ncols</span><span class="p">),</span> <span class="n">ncols</span><span class="p">,</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">cdplt</span><span class="o">.</span><span class="n">scatter_gene</span><span class="p">(</span>
        <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
        <span class="n">x</span><span class="o">=</span><span class="s1">&#39;splice&#39;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s1">&#39;unsplice&#39;</span><span class="p">,</span>
        <span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cellDancer_df</span><span class="p">,</span>
        <span class="n">custom_xlim</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">custom_ylim</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">colors</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_erythroid</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
        <span class="n">s</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">velocity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">gene</span><span class="o">=</span><span class="n">gene_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">gene_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_gastrulation_13_0.png" src="../_images/notebooks_case_study_gastrulation_13_0.png" />
</div>
</div>
</div>
<div class="section" id="Project-the-RNA-velocity-onto-the-embedding-space">
<h2>Project the RNA velocity onto the embedding space<a class="headerlink" href="#Project-the-RNA-velocity-onto-the-embedding-space" title="Permalink to this headline">¶</a></h2>
<p>To project the prediction of RNA velocity onto the embedding space and to estimate pseudotime using all 2,000 genes, predicted result can be downloaded from <a class="reference external" href="https://drive.google.com/file/d/1J4GB-Bn2SdmK56wXka5TdvkImspPwIjH/view?usp=sharing">GastrulationErythroid_cellDancer_estimation.csv.zip</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load the prediction result of all genes</span>
<span class="n">cellDancer_df_path</span> <span class="o">=</span> <span class="s1">&#39;your_path/GastrulationErythroid_cellDancer_estimation.csv&#39;</span>
<span class="n">cellDancer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cellDancer_df_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We calculate the projection of RNA velocity on the embedding with <code class="docutils literal notranslate"><span class="pre">cd.compute_cell_velocity()</span></code>. The projected direction on embedding space, i.e. columns ‘velocity1’ and ‘velocity2’ are added to the original dataframe. We use <code class="docutils literal notranslate"><span class="pre">cdplt.scatter_cell()</span></code> to display the predicted direction on embedding space.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute cell velocity</span>
<span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">compute_cell_velocity</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">projection_neighbor_choice</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="n">expression_scale</span><span class="o">=</span><span class="s1">&#39;power10&#39;</span><span class="p">,</span> <span class="n">projection_neighbor_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">speed_up</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span><span class="mi">100</span><span class="p">))</span>

<span class="c1"># plot cell velocity</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
<span class="n">cdplt</span><span class="o">.</span><span class="n">scatter_cell</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span>
                   <span class="n">cellDancer_df</span><span class="p">,</span>
                   <span class="n">colors</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_erythroid</span><span class="p">,</span>
                   <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
                   <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                   <span class="n">velocity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                   <span class="n">legend</span><span class="o">=</span><span class="s1">&#39;on&#39;</span><span class="p">,</span>
                   <span class="n">min_mass</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
                   <span class="n">arrow_grid</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span>
                   <span class="n">custom_xlim</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="mi">6</span><span class="p">,</span><span class="mi">13</span><span class="p">],</span>
                   <span class="n">custom_ylim</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">16</span><span class="p">],</span> <span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_gastrulation_18_0.png" src="../_images/notebooks_case_study_gastrulation_18_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cellDancer_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cellIndex</th>
      <th>gene_name</th>
      <th>unsplice</th>
      <th>splice</th>
      <th>unsplice_predict</th>
      <th>splice_predict</th>
      <th>alpha</th>
      <th>beta</th>
      <th>gamma</th>
      <th>loss</th>
      <th>cellID</th>
      <th>clusters</th>
      <th>embedding1</th>
      <th>embedding2</th>
      <th>index</th>
      <th>velocity1</th>
      <th>velocity2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>Ift81</td>
      <td>0.010658</td>
      <td>0.026321</td>
      <td>0.009261</td>
      <td>0.029549</td>
      <td>0.023595</td>
      <td>0.042598</td>
      <td>0.073999</td>
      <td>0.039626</td>
      <td>cell_363</td>
      <td>Blood progenitors 2</td>
      <td>3.460521</td>
      <td>15.574629</td>
      <td>0</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>Ift81</td>
      <td>0.000000</td>
      <td>0.044266</td>
      <td>0.000946</td>
      <td>0.037888</td>
      <td>0.020649</td>
      <td>0.042942</td>
      <td>0.074502</td>
      <td>0.039626</td>
      <td>cell_382</td>
      <td>Blood progenitors 2</td>
      <td>2.490433</td>
      <td>14.971734</td>
      <td>1</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>Ift81</td>
      <td>0.000000</td>
      <td>0.064559</td>
      <td>0.000885</td>
      <td>0.055191</td>
      <td>0.019326</td>
      <td>0.042876</td>
      <td>0.075031</td>
      <td>0.039626</td>
      <td>cell_385</td>
      <td>Blood progenitors 2</td>
      <td>2.351203</td>
      <td>15.267069</td>
      <td>2</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>Ift81</td>
      <td>0.000000</td>
      <td>0.020756</td>
      <td>0.001014</td>
      <td>0.017791</td>
      <td>0.022149</td>
      <td>0.043030</td>
      <td>0.073879</td>
      <td>0.039626</td>
      <td>cell_393</td>
      <td>Blood progenitors 2</td>
      <td>5.899098</td>
      <td>14.388825</td>
      <td>3</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>Ift81</td>
      <td>0.000000</td>
      <td>0.013184</td>
      <td>0.001037</td>
      <td>0.011305</td>
      <td>0.022633</td>
      <td>0.043055</td>
      <td>0.073676</td>
      <td>0.039626</td>
      <td>cell_398</td>
      <td>Blood progenitors 2</td>
      <td>4.823139</td>
      <td>15.374831</td>
      <td>4</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>24657995</th>
      <td>12324</td>
      <td>Mcrip1</td>
      <td>0.000000</td>
      <td>1.128435</td>
      <td>0.000157</td>
      <td>1.125867</td>
      <td>0.005920</td>
      <td>0.038023</td>
      <td>0.013131</td>
      <td>0.051755</td>
      <td>cell_139318</td>
      <td>Erythroid3</td>
      <td>8.032358</td>
      <td>7.603037</td>
      <td>12324</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>24657996</th>
      <td>12325</td>
      <td>Mcrip1</td>
      <td>0.024356</td>
      <td>0.970672</td>
      <td>0.016090</td>
      <td>1.428982</td>
      <td>0.008338</td>
      <td>0.037053</td>
      <td>0.013950</td>
      <td>0.051755</td>
      <td>cell_139321</td>
      <td>Erythroid3</td>
      <td>10.352904</td>
      <td>6.446736</td>
      <td>12325</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>24657997</th>
      <td>12326</td>
      <td>Mcrip1</td>
      <td>0.000000</td>
      <td>0.899107</td>
      <td>0.000175</td>
      <td>0.897000</td>
      <td>0.006575</td>
      <td>0.037644</td>
      <td>0.013522</td>
      <td>0.051755</td>
      <td>cell_139326</td>
      <td>Erythroid3</td>
      <td>9.464873</td>
      <td>7.261099</td>
      <td>12326</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>24657998</th>
      <td>12327</td>
      <td>Mcrip1</td>
      <td>0.017375</td>
      <td>1.398107</td>
      <td>0.011387</td>
      <td>1.729827</td>
      <td>0.006885</td>
      <td>0.037765</td>
      <td>0.013271</td>
      <td>0.051755</td>
      <td>cell_139327</td>
      <td>Erythroid3</td>
      <td>9.990495</td>
      <td>7.243880</td>
      <td>12327</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>24657999</th>
      <td>12328</td>
      <td>Mcrip1</td>
      <td>0.005194</td>
      <td>0.988933</td>
      <td>0.003540</td>
      <td>1.086199</td>
      <td>0.006822</td>
      <td>0.037569</td>
      <td>0.013596</td>
      <td>0.051755</td>
      <td>cell_139330</td>
      <td>Erythroid3</td>
      <td>8.260699</td>
      <td>7.935455</td>
      <td>12328</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>24658000 rows × 17 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Estimate-pseudotime">
<h2>Estimate pseudotime<a class="headerlink" href="#Estimate-pseudotime" title="Permalink to this headline">¶</a></h2>
<p>Based on the projection of RNA velocity on embedding space, we estimate the pseudotime with <code class="docutils literal notranslate"><span class="pre">cd.pseudo_time()</span></code>.</p>
<p>The long trajectories used for pseudotime dertermination are (optionally) shown. In this case, the three long trajectories are colored from light to dark according to their own unadjusted pseudotime. The cells are colored according to the long trajectories that are used for determination of the pseudotime of the cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">random</span>
<span class="c1"># set parameters</span>
<span class="n">dt</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="n">t_total</span> <span class="o">=</span> <span class="p">{</span><span class="n">dt</span><span class="p">:</span><span class="nb">int</span><span class="p">(</span><span class="mi">10</span><span class="o">/</span><span class="n">dt</span><span class="p">)}</span>
<span class="n">n_repeats</span> <span class="o">=</span> <span class="mi">10</span>

<span class="c1"># estimate pseudotime</span>
<span class="n">cellDancer_df</span> <span class="o">=</span> <span class="n">cd</span><span class="o">.</span><span class="n">pseudo_time</span><span class="p">(</span><span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cellDancer_df</span><span class="p">,</span>
                               <span class="n">grid</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">30</span><span class="p">),</span>
                               <span class="n">dt</span><span class="o">=</span><span class="n">dt</span><span class="p">,</span>
                               <span class="n">t_total</span><span class="o">=</span><span class="n">t_total</span><span class="p">[</span><span class="n">dt</span><span class="p">],</span>
                               <span class="n">n_repeats</span><span class="o">=</span><span class="n">n_repeats</span><span class="p">,</span>
                               <span class="n">speed_up</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span><span class="mi">100</span><span class="p">),</span>
                               <span class="n">n_paths</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
                               <span class="n">plot_long_trajs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                               <span class="n">psrng_seeds_diffusion</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_repeats</span><span class="p">)],</span>
                               <span class="n">n_jobs</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Pseudo random number generator seeds are set to:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Generating Trajectories: 100%|<span class="ansi-blue-fg">████████████</span>| 9510/9510 [00:02&lt;00:00, 3175.15it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_gastrulation_23_2.png" src="../_images/notebooks_case_study_gastrulation_23_2.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
There are 3 clusters.
[0 1 2]
--- 257.67111110687256 seconds ---
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot pseudotime</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="n">im</span><span class="o">=</span><span class="n">cdplt</span><span class="o">.</span><span class="n">scatter_cell</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span><span class="n">cellDancer_df</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="s1">&#39;pseudotime&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">velocity</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">custom_xlim</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="mi">11</span><span class="p">),</span> <span class="n">custom_ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">18</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_gastrulation_24_0.png" src="../_images/notebooks_case_study_gastrulation_24_0.png" />
</div>
</div>
<p>The connection network below is another method to display pseudotime (<code class="docutils literal notranslate"><span class="pre">cdplt.PTO_Graph</span></code>). The edge lengths indicate the time difference between nodes (the closer in pseudotime, the shorter the edge length). The sizes of the nodes are proportional to the pseudotime.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>

<span class="n">cdplt</span><span class="o">.</span><span class="n">PTO_Graph</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span>
                <span class="n">cellDancer_df</span><span class="p">,</span>
                <span class="n">node_layout</span><span class="o">=</span><span class="s1">&#39;forcedirected&#39;</span><span class="p">,</span>
                <span class="n">PRNG_SEED</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                <span class="n">use_edge_bundling</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">node_colors</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_erythroid</span><span class="p">,</span>
                <span class="n">edge_length</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                <span class="n">node_sizes</span><span class="o">=</span><span class="s1">&#39;pseudotime&#39;</span><span class="p">,</span>
                <span class="n">colorbar</span><span class="o">=</span><span class="s1">&#39;on&#39;</span><span class="p">,</span>
                <span class="n">legend</span><span class="o">=</span><span class="s1">&#39;on&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_gastrulation_26_1.png" src="../_images/notebooks_case_study_gastrulation_26_1.png" />
</div>
</div>
</div>
<div class="section" id="Display-the-abundance-of-spliced-RNA-along-pseudotime">
<h2>Display the abundance of spliced RNA along pseudotime<a class="headerlink" href="#Display-the-abundance-of-spliced-RNA-along-pseudotime" title="Permalink to this headline">¶</a></h2>
<p>We visualize the spliced RNA abundance of genes along pseudotime with <code class="docutils literal notranslate"><span class="pre">cdplt.scatter_gene()</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ncols</span><span class="o">=</span><span class="mi">5</span>
<span class="n">height</span><span class="o">=</span><span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="o">/</span><span class="n">ncols</span><span class="p">)</span><span class="o">*</span><span class="mi">4</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="n">height</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="o">/</span><span class="n">ncols</span><span class="p">),</span> <span class="n">ncols</span><span class="p">,</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">cdplt</span><span class="o">.</span><span class="n">scatter_gene</span><span class="p">(</span>
        <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
        <span class="n">x</span><span class="o">=</span><span class="s1">&#39;pseudotime&#39;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s1">&#39;splice&#39;</span><span class="p">,</span>
        <span class="n">cellDancer_df</span><span class="o">=</span><span class="n">cellDancer_df</span><span class="p">,</span>
        <span class="n">custom_xlim</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">custom_ylim</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">colors</span><span class="o">=</span><span class="n">colormap</span><span class="o">.</span><span class="n">colormap_erythroid</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
        <span class="n">s</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">velocity</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">gene</span><span class="o">=</span><span class="n">gene_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">gene_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_case_study_gastrulation_29_0.png" src="../_images/notebooks_case_study_gastrulation_29_0.png" />
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="case_study_neuro.html" class="btn btn-neutral float-right" title="Inferring RNA velocities on each branch for branching genes in the mouse hippocampus development" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../data_preprocessing.html" class="btn btn-neutral float-left" title="Data Preparation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, 6.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>