

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Data Preparation &mdash; cellDancer  documentation</title>
  

  
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/jquery.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/underscore.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
    
    <script type="text/javascript" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation" href="notebooks/case_study_gastrulation.html" />
    <link rel="prev" title="Installation" href="notebooks/installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> cellDancer
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">cellDancer</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="API.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="release_notes.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="notebooks/installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Data Preparation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#transfer-adata-format-to-dataframe">Transfer adata format to dataframe</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_gastrulation.html">Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_neuro.html">Inferring RNA velocities on each branch for branching genes in the mouse hippocampus development</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_pancreas.html">Deciphering cell identity using the cell-specific reaction rates in pancreatic endocrinogenesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_hgforebrian.html">Decoding the human embryo glutamatergic neurogenesis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application of dynamo</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_pancreas_dynamo.html">Deciphering gene regulation through vector fields analysis in pancreatic endocrinogenesis</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">cellDancer</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Data Preparation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/data_preprocessing.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="data-preparation">
<h1>Data Preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">Â¶</a></h1>
<p>In our tutorials of <a class="reference external" href="notebooks/case_study_gastrulation.html">Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation data</a>, <a class="reference external" href="notebooks/case_study_neuro.html">Inferring RNA velocities on each branch for branching genes in the mouse hippocampus development</a>, <a class="reference external" href="notebooks/case_study_pancreas.html">Deciphering cell identity using the cell-specific reaction rates in pancreatic endocrinogenesis</a>, and <a class="reference external" href="notebooks/case_study_hgforebrian.html">Decoding the human embryo glutamatergic neurogenesis</a>, the data has already been prepared. To load your own data, the data format and the method to prepare the data is introduced below.</p>
<p>The input data for the prediction of velocity is in <code class="docutils literal notranslate"><span class="pre">csv</span></code> format. And could be loaded using <code class="docutils literal notranslate"><span class="pre">pd.read_csv()</span></code>. Taking the <a class="reference external" href="https://drive.google.com/file/d/1XYWaJ3AyMspW7Wfy43LAcgsYm9GGDDh1/view?usp=sharing">cell_type_u_s_sample_df.csv</a> below as an example of 5 cells corresponding to 2 genes as shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">cell_type_u_s</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;your_path/cell_type_u_s_sample_df.csv&#39;</span><span class="p">)</span>
<span class="n">cell_type_u_s</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_images/cell_type_u_s_sample_df.png"><img alt="cell_type_u_s_sample_df" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_images/cell_type_u_s_sample_df.png" style="width: 100%;" /></a>
<p>The gene information is represented by columns of <code class="docutils literal notranslate"><span class="pre">gene_name</span></code>, <code class="docutils literal notranslate"><span class="pre">unsplice</span></code>, and <code class="docutils literal notranslate"><span class="pre">splice</span></code>. The cell information is represented by columns <code class="docutils literal notranslate"><span class="pre">cellID</span></code>, <code class="docutils literal notranslate"><span class="pre">clusters</span></code>, <code class="docutils literal notranslate"><span class="pre">embedding1</span></code>, and <code class="docutils literal notranslate"><span class="pre">embedding2</span></code>. <code class="docutils literal notranslate"><span class="pre">unsplice</span></code> and <code class="docutils literal notranslate"><span class="pre">splice</span></code> columns represent the spliced and unspliced counts, separately. <code class="docutils literal notranslate"><span class="pre">cellID</span></code> is the unique id of each cell. <code class="docutils literal notranslate"><span class="pre">clusters</span></code> represent the cell type of each cell. <code class="docutils literal notranslate"><span class="pre">embedding1</span></code> and <code class="docutils literal notranslate"><span class="pre">embedding2</span></code> are the 2-dimensional representation of all cells such as UMAP, PCA, or t-SNE. The sample could be viewed by</p>
<div class="section" id="transfer-adata-format-to-dataframe">
<h2>Transfer adata format to dataframe<a class="headerlink" href="#transfer-adata-format-to-dataframe" title="Permalink to this headline">Â¶</a></h2>
<p>The two count matrices of unspliced and spliced abundances could be obtained from standard sequencing protocols. They could be counted by <a class="reference external" href="http://velocyto.org/velocyto.py/tutorial/cli.html#running-velocyto">velocyto</a> or <a class="reference external" href="https://linnarssonlab.org/loompy/kallisto/index.html">loompy/kallisto</a> pipeline.</p>
<p>We also provide a function (<code class="docutils literal notranslate"><span class="pre">cd.adata_to_df_with_embed()</span></code>) to transfer from <a class="reference external" href="https://anndata-tutorials.readthedocs.io/en/latest/getting-started.html">Anndata</a> (a format of storing annotated data, usually are loom file) to <code class="docutils literal notranslate"><span class="pre">csv</span></code> format. For example, after the <a class="reference external" href="https://scvelo.readthedocs.io/VelocityBasics/">preprocessing in Anndata format</a>. To transfer to pandas.DataFrame/csv format, <code class="docutils literal notranslate"><span class="pre">cd.adata_to_df_with_embed()</span></code> could be used. For example, in the command of:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">celldancer.utilities</span> <span class="k">as</span> <span class="nn">cdutil</span>
<span class="n">cdutil</span><span class="o">.</span><span class="n">adata_to_df_with_embed</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                              <span class="n">us_para</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Mu&#39;</span><span class="p">,</span><span class="s1">&#39;Ms&#39;</span><span class="p">],</span>
                              <span class="n">cell_type_para</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span>
                              <span class="n">embed_para</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span>
                              <span class="n">save_path</span><span class="o">=</span><span class="s1">&#39;cell_type_u_s.csv&#39;</span><span class="p">,</span>
                              <span class="n">gene_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Hba-x&#39;</span><span class="p">,</span><span class="s1">&#39;Smim1&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">splice</span></code> and <code class="docutils literal notranslate"><span class="pre">unsplice</span></code> columns are obtained from the <code class="docutils literal notranslate"><span class="pre">['Ms',</span> <span class="pre">'Mu']</span></code> attributes of <code class="docutils literal notranslate"><span class="pre">adata.layers</span></code>. Also, <code class="docutils literal notranslate"><span class="pre">cellID</span></code> column is obtained from <code class="docutils literal notranslate"><span class="pre">adata.obs.index</span></code>. <code class="docutils literal notranslate"><span class="pre">clusters</span></code> column is obtained from <code class="docutils literal notranslate"><span class="pre">['celltype']</span></code> of <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>. The <code class="docutils literal notranslate"><span class="pre">embedding1</span></code> and <code class="docutils literal notranslate"><span class="pre">embedding2</span></code> columns are obtained from <code class="docutils literal notranslate"><span class="pre">['X_umap']</span></code> attribute of <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code>.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="notebooks/case_study_gastrulation.html" class="btn btn-neutral float-right" title="Delineating the transcriptional boost on single-cell resolution in gastrulation erythroid maturation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="notebooks/installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, 6.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>