

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Frequently asked questions &mdash; cellDancer  documentation</title>
  

  
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/documentation_options.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/jquery.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/underscore.js"></script>
        <script src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    
    <script type="text/javascript" src="https://raw.githack.com/GuangyuWangLab2021/cellDancer_website/main/docs/_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="notebooks/installation.html" />
    <link rel="prev" title="Release Notes" href="release_notes.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> cellDancer
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">cellDancer</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About cellDancer</a></li>
<li class="toctree-l1"><a class="reference internal" href="API.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="release_notes.html">Release Notes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Frequently asked questions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="data_preprocessing.html">Data Preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_gastrulation.html">Case study 1: Gastrulation erythroid maturation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_neuro.html">Case Study 2: Mouse hippocampus development</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_pancreas.html">Case study 3: Pancreatic endocrinogenesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_hgforebrian.html">Case study 4: Human embryo glutamatergic neurogenesis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application of dynamo</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks/case_study_pancreas_dynamo.html">Downstream analysis using Dynamo</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">cellDancer</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Frequently asked questions</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/FAQ.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="frequently-asked-questions">
<h1>Frequently asked questions<a class="headerlink" href="#frequently-asked-questions" title="Permalink to this headline">¶</a></h1>
<p>Below are the frequently asked questions we’ve received. If you have any other question related to your specific contition, welcome to post it in our <a class="reference external" href="https://github.com/GuangyuWangLab2021/cellDancer/issues">github issue page</a> or email to <a class="reference external" href="mailto:sli5&#37;&#52;&#48;houstonmethodist&#46;org">sli5<span>&#64;</span>houstonmethodist<span>&#46;</span>org</a>.</p>
<div class="section" id="questions-of-transferring-adata-format-to-dataframe">
<h2>Questions of transferring adata format to dataframe.<a class="headerlink" href="#questions-of-transferring-adata-format-to-dataframe" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://guangyuwanglab2021.github.io/cellDancer_website/data_preprocessing.html">Data Preprocessing</a> contains the introduction for analyzing with you own data.</p>
<p><strong>Q1</strong>: When transferring adata format to dataframe, how to get all the genes in the adata?</p>
<p><strong>A1</strong>:Not inputting the gene_list will let cdutil.cdadata_to_df_with_embed function get all genes by default. Details could be found in our <a class="reference external" href="https://github.com/GuangyuWangLab2021/cellDancer/issues/4">Github issue page</a>.</p>
<p><strong>Q2</strong>: How can we know which gene to use? Where are these genes in case studies come from? Do we need to search for the branching gene first?</p>
<p><strong>A2</strong>: In most cases, all high variable genes could be used as mentioned in <a class="reference external" href="https://guangyuwanglab2021.github.io/cellDancer_website/data_preprocessing.html">Data Preprocessing</a> with <code class="docutils literal notranslate"><span class="pre">scv.pp.filter_and_normalize(adata,min_shared_counts=20,n_top_genes=2000)</span></code>. We manually searched the branching genes and combined them with the reported branching gene list. If you want to study branching genes, you may also manually search for branching genes.</p>
<p><strong>Q3</strong>: When using the adata_to_df_with_embed, I got an error: ValueError: If using all scalar values, you must pass an index. Below is how my adata looks like</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 23496 × 18964
    obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;donor_id&#39;, &#39;prob_max&#39;, &#39;prob_doublet&#39;, &#39;n_vars&#39;, &#39;best_singlet&#39;, &#39;doublet_logLikRatio&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;pANN&#39;, &#39;gender&#39;, &#39;location&#39;, &#39;donors&#39;, &#39;SCT_snn_res.0.5&#39;, &#39;seurat_clusters&#39;, &#39;S.Score&#39;, &#39;G2M.Score&#39;, &#39;Phase&#39;, &#39;old.ident&#39;, &#39;RNA_snn_res.0.4&#39;, &#39;cell_types&#39;, &#39;batch&#39;, &#39;initial_size_unspliced&#39;, &#39;initial_size_spliced&#39;, &#39;initial_size&#39;
    var: &#39;name&#39;, &#39;Accession&#39;, &#39;Chromosome&#39;, &#39;End&#39;, &#39;Start&#39;, &#39;Strand&#39;
    uns: &#39;cell_types_colors&#39;
    obsm: &#39;X_harmony&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;ambiguous&#39;, &#39;matrix&#39;, &#39;spliced&#39;, &#39;unspliced&#39;
</pre></div>
</div>
<p><strong>A3</strong>: Seems like you don’t have ‘Mu’ and ‘Ms’ in ‘ layers’ of adata. The codes below will help you to get the ‘Mu’ and ‘Ms’ in your ‘layers’ of adata.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># !pip install scvelo --upgrade --quiet</span>
<span class="kn">import</span> <span class="nn">scvelo</span> <span class="k">as</span> <span class="nn">scv</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_and_normalize</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_shared_counts</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">moments</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span> <span class="c1"># cell amount will influence the setting of n_neighbors</span>
</pre></div>
</div>
<p>Details could be found in our <a class="reference external" href="https://github.com/GuangyuWangLab2021/cellDancer/issues/4">Github issue page</a>.</p>
</div>
<div class="section" id="errors-met-when-following-the-case-study-on-own-data">
<h2>Errors met when following the case study on own data.<a class="headerlink" href="#errors-met-when-following-the-case-study-on-own-data" title="Permalink to this headline">¶</a></h2>
<p><strong>Q1</strong>. I am using my own data, why my prediction result is showing the arrow and cell type symbol only when using ‘cdplt.scatter_gene’ and ‘cdplt.scatter_cell’. Also, where does colormap.colormap_neuro come from?</p>
<p><strong>A1</strong>: Your prediction result is only showing the arrow and cell type symbol only, this is because you may need to build your own colormap. colormap.colormap_neuro is the internal customized colormap dictionary we used in celldancer for our specific color set. If you want to color your own celltype, there are two ways to obtain the color map.</p>
<p>1) Build colormap using the default color style of cellDancer.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">celldancer</span> <span class="k">as</span> <span class="nn">cd</span>
<span class="n">Import</span> <span class="n">celldancer</span><span class="o">.</span><span class="n">cdplt</span> <span class="k">as</span> <span class="n">cdplt</span>
<span class="n">Import</span> <span class="n">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="n">cell_type_u_s</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;./cell_type_u_s.csv&#39;</span><span class="p">)</span> <span class="c1"># transferred pandas DataFrame from adata</span>
<span class="n">lst_of_celltype</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">cell_type_u_s</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">())</span>
<span class="n">colormap1</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">plotting</span><span class="o">.</span><span class="n">colormap</span><span class="o">.</span><span class="n">build_colormap</span><span class="p">(</span><span class="n">lst_of_celltype</span><span class="p">)</span>
</pre></div>
</div>
<p>Then, you can set <code class="docutils literal notranslate"><span class="pre">colors=colormap1</span></code> when using <code class="docutils literal notranslate"><span class="pre">cdplt.scatter_gene()</span></code> and <code class="docutils literal notranslate"><span class="pre">cdplt.scatter_cell()</span></code> functions.</p>
<p>2) Customize own color style.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>import celldancer.cdplt as cdplt
colormap2=dict()
colormap2[&#39;celltype1&#39;]=&#39;#3B9AB2&#39; # the Hex code of color
colormap2[&#39;celltype2&#39;]=&#39;#710162&#39;
…# manually set the color for each celltype
</pre></div>
</div>
<p>Then, you can set <code class="docutils literal notranslate"><span class="pre">colors=colormap2</span></code> when using <code class="docutils literal notranslate"><span class="pre">cdplt.scatter_gene()</span></code> and <code class="docutils literal notranslate"><span class="pre">cdplt.scatter_cell()</span></code> functions.</p>
<p><strong>Q2</strong>: I was wondering if there was a way to cap the memory usage?</p>
<p><strong>A2</strong>: You can start with <code class="docutils literal notranslate"><span class="pre">n_jobs=1</span></code> in <a href="#id4"><span class="problematic" id="id5">``</span></a>cd.velocity `` to see if it works or not. Details could be found in our <a class="reference external" href="https://github.com/GuangyuWangLab2021/cellDancer/issues/7">Github issue page</a>.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="notebooks/installation.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="release_notes.html" class="btn btn-neutral float-left" title="Release Notes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright June, 2022.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>